File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        // Build unordered_map for O(1) key lookups
        unordered_map<string, string> map;
        for (const auto& pair : knowledge) {
            map[pair[0]] = pair[1];
        }

        string result = "";
        string key = "";
        bool inBracket = false;

        for (char c : s) {
            if (c == '(') {
                inBracket = true;
            } else if (c == ')') {
                inBracket = false;
                auto it = map.find(key);
                if (it != map.end()) {
                    result += it->second;
                } else {
                    result += '?';
                }
                key.clear(); // Reset key buffer
            } else {
                if (inBracket) {
                    key += c;
                } else {
                    result += c;
                }
            }
        }

        return result;
    }
};

File: 1934-evaluate-the-bracket-pairs-of-a-string/evaluate-the-bracket-pairs-of-a-string.java
import java.util.*;

class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        // Build HashMap for O(1) key lookups
        Map<String, String> map = new HashMap<>();
        for (List<String> pair : knowledge) {
            map.put(pair.get(0), pair.get(1));
        }

        StringBuilder result = new StringBuilder();
        StringBuilder key = new StringBuilder();
        boolean inBracket = false;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == '(') {
                inBracket = true;
            } else if (c == ')') {
                inBracket = false;
                result.append(map.getOrDefault(key.toString(), "?"));
                key.setLength(0); // Reset key buffer
            } else {
                if (inBracket) {
                    key.append(c);
                } else {
                    result.append(c);
                }
            }
        }

        return result.toString();
    }
}